apps/vscode/command_client/vscode.py

```diff
@@ -0,0 +1,92 @@
+from pickle import FALSE
+from typing import Any
+
+from talon import Context, Module, actions
+
+from .command_client import NoFileServerException, NotSet, run_command
+
+ctx = Context()
+
+ctx.matches = r"""
+app: vscode
+"""
+ctx.tags = ["user.command_client"]
+mod = Module()
+
+
+def command_server_or_client_fallback(command_id: str, wait: bool):
+    """Execute command via command server and us command pallete if directory not present."""
+    try:
+        run_command(command_id, wait_for_finish=wait)
+    except NoFileServerException:
+        actions.user.command_palette()
+        actions.user.paste(command_id)
+        actions.key("enter")
+        print(
+            "Command issues via command palette for better performance install the VSCode extension for Talon"
+        )
+
+
+@ctx.action_class("user")
+class VsCodeAction:
+    def command_server_directory() -> str:
+        return "vscode-command-server"
+
+
+# These commands are shims, to provide backwards compatibility, they may be removed in the fuuture.
+# Prefer the run_command... version in command_client.
+@mod.action_class
+class Actions:
```

Comment:
What happened to the pre-phrase signal code? Unless I've missed something, it appears to have been dropped

Reply:
Ah nvm I see it now. A bit confusing how it's declared in multiple places

What's the point of allowing applications to turn it off? Seems simpler to just leave it on and let all applications share the same impl

Reply:
See command client 368 its in ctx which uses the match belwo to check we are in the correct context. I put a comment in command_client_tag to explain when to overwrite

```Python
ctx.matches = r"""
tag: command_client
"""
```

Reply:
I'll remove it from the actions so that it cannot be turned of.

Reply:
Done

Reply:
Uh, this seems to have been changed into a non-working state. There's no module which declares `emit_pre_phrase_signal`, only a context which implements it, which should be producing an error in the talon log. We probably want the default implementation to be `return False`. There's no sense touching a file if we're not in a command client. I'll go change this now.

Reply:
Ok, this should be working now. @pokey is this okay? Did you want us to _always_ touch the pre-phrase file, even if we don't have a command-client app focused?

Reply:
interestingly, it doesn't seem to be causing an error in the talon log. not sure why.

Reply:
ok. it does error if I actually try to use the action, though. I'll try to minimize that and report it as a talon bug.

Reply:
ok, I've reported the underlying bug to talon, https://github.com/talonvoice/talon/issues/525.